### ISSUE TITLE
Modules that are called during a trace must be registered as submodules of the thing being traced.`

### ISSUE THREAD

**Issue by Dakkay** [Xilinx/brevitas#174]
Hey everybody i am getting the following mistake and I have little understanding of what i can do to solve it.

<__torch__.brevitas.core.quant.RescalingIntQuant object at 0x55b84e4f5130>
but it is not part of the active trace. Modules that are called during
a trace must be registered as submodules of the thing being traced.`

I am trying to do a Onxx-export in order to implement brevitas on FINN.

Is this any known problem that is currently being resolved or am I doing something wrong ?

Thanks

**Comment by volcacius** [comment]
Hello,

What kind of model are you trying to export? Without more details I can't help you.

Alessandro

**Comment by Dakkay** [comment]
`"""
quantlenet implemented as brevitas example
"""
import torch
import torch.nn as nn
import torch.nn.functional as F
import brevitas.nn as qnn
from brevitas.core.quant import QuantType
from torch.nn import Module


class QuantLeNet(Module):
    def __init__(self, vocab_size, embedding_dim, kernel_sizes, num_filters, num_classes, d_prob, mode, TEXT):
        super(QuantLeNet, self).__init__()
        self.vocab_size = vocab_size
        self.embedding_dim = embedding_dim
        self.kernel_sizes = kernel_sizes
        self.num_filters = num_filters
        self.num_classes = num_classes
        self.d_prob = d_prob
        self.mode = mode
        self.embedding = nn.Embedding(vocab_size, embedding_dim, padding_idx=1)
        self.load_embeddings(TEXT)

        # is this ok? if not split list comprehension. it just holds them anyway,
        # no calculations involved
        self.conv = nn.ModuleList([
            qnn.QuantConv1d(
                in_channels=embedding_dim,
                out_channels=num_filters,
                kernel_size=k, stride=1,
                weight_quant_type=QuantType.INT,
                weight_bit_width=12) for k in kernel_sizes])

        # next line is iffy. am i allowed? just remove if i cant, it helps
        # with regularization but maybe we can find another way.
        # at the same time, it *should* be REALLY easy to implement
        self.dropout = nn.Dropout(d_prob)
        self.fc = qnn.QuantLinear(len(kernel_sizes) * num_filters, num_classes,
                                  bias=True,
                                  weight_quant_type=QuantType.INT,
                                  weight_bit_width=32)
        """
        self.conv1 = qnn.QuantConv2d(3, 6, 5,
                                     weight_quant_type=QuantType.INT,
                                     weight_bit_width=8)
        self.relu1 = qnn.QuantReLU(
            quant_type=QuantType.INT, bit_width=8, max_val=6)
        self.conv2 = qnn.QuantConv2d(6, 16, 5,
                                     weight_quant_type=QuantType.INT,
                                     weight_bit_width=8)
        self.relu2 = qnn.QuantReLU(
            quant_type=QuantType.INT, bit_width=8, max_val=6)
        self.fc1 = qnn.QuantLinear(16*5*5, 120, bias=True,
                                   weight_quant_type=QuantType.INT,
                                   weight_bit_width=8)
        self.relu3 = qnn.QuantReLU(
            quant_type=QuantType.INT, bit_width=8, max_val=6)
        self.fc2 = qnn.QuantLinear(120, 84, bias=True,
                                   weight_quant_type=QuantType.INT,
                                   weight_bit_width=8)
        self.relu4 = qnn.QuantReLU(
            quant_type=QuantType.INT, bit_width=8, max_val=6)
        self.fc3 = qnn.QuantLinear(84, 10, bias=False,
                                   weight_quant_type=QuantType.INT,
                                   weight_bit_width=8)
        """

    def forward(self, x):
        batch_size, sequence_length = x.shape
        relu = qnn.QuantReLU(
            quant_type=QuantType.INT, bit_width=32, max_val=6)
        sigmoid = qnn.QuantSigmoid(
            quant_type=QuantType.INT, bit_width=32)

        x = self.embedding(x).transpose(1, 2)
        x = [relu(conv(x)) for conv in self.conv]
        x = [F.max_pool1d(c, c.size(-1)).squeeze(dim=-1) for c in x]
        x = torch.cat(x, dim=1)
        x = self.fc(self.dropout(x))
        x = sigmoid(x)
        return x.squeeze()

    def load_embeddings(self, TEXT):
        if 'static' in self.mode:
            self.embedding.weight.data.copy_(TEXT.vocab.vectors)
            if 'non' not in self.mode:
                self.embedding.weight.data.requires_grad = False
                print('Loaded pretrained embeddings, weights are not trainable.')
            else:
                self.embedding.weight.data.requires_grad = True
                print('Loaded pretrained embeddings, weights are trainable.')
        elif self.mode == 'rand':
            print('Randomly initialized embeddings are used.')
        else:
            raise ValueError(
                'Unexpected value of mode. Please choose from static, nonstatic, rand.')`

**Comment by Dakkay** [comment]
Here is the code 
Mainly taken from examples on the brevitas repo.

`"""
quantlenet implemented as brevitas example
"""

class QuantLeNet(Module):
    def __init__(self, vocab_size, embedding_dim, kernel_sizes, num_filters, num_classes, d_prob, mode, TEXT):
        super(QuantLeNet, self).__init__()
        self.vocab_size = vocab_size
        self.embedding_dim = embedding_dim
        self.kernel_sizes = kernel_sizes
        self.num_filters = num_filters
        self.num_classes = num_classes
        self.d_prob = d_prob
        self.mode = mode
        self.embedding = nn.Embedding(vocab_size, embedding_dim, padding_idx=1)
        self.load_embeddings(TEXT)

  
        self.conv = nn.ModuleList([
            qnn.QuantConv1d(
                in_channels=embedding_dim,
                out_channels=num_filters,
                kernel_size=k, stride=1,
                weight_quant_type=QuantType.INT,
                weight_bit_width=12) for k in kernel_sizes])

        self.dropout = nn.Dropout(d_prob)
        self.fc = qnn.QuantLinear(len(kernel_sizes) * num_filters, num_classes,
                                  bias=True,
                                  weight_quant_type=QuantType.INT,
                                  weight_bit_width=32)
   

    def forward(self, x):
        batch_size, sequence_length = x.shape
        relu = qnn.QuantReLU(
            quant_type=QuantType.INT, bit_width=32, max_val=6)
        sigmoid = qnn.QuantSigmoid(
            quant_type=QuantType.INT, bit_width=32)

        x = self.embedding(x).transpose(1, 2)
        x = [relu(conv(x)) for conv in self.conv]
        x = [F.max_pool1d(c, c.size(-1)).squeeze(dim=-1) for c in x]
        x = torch.cat(x, dim=1)
        x = self.fc(self.dropout(x))
        x = sigmoid(x)
        return x.squeeze()

    def load_embeddings(self, TEXT):
        if 'static' in self.mode:
            self.embedding.weight.data.copy_(TEXT.vocab.vectors)
            if 'non' not in self.mode:
                self.embedding.weight.data.requires_grad = False
                print('Loaded pretrained embeddings, weights are not trainable.')
            else:
                self.embedding.weight.data.requires_grad = True
                print('Loaded pretrained embeddings, weights are trainable.')
        elif self.mode == 'rand':
            print('Randomly initialized embeddings are used.')
        else:
            raise ValueError(
                'Unexpected value of mode. Please choose from static, nonstatic, rand.')`

**Comment by Dakkay** [comment]
`"""
quantlenet implemented as brevitas example
"""

class QuantLeNet(Module):
    def __init__(self, vocab_size, embedding_dim, kernel_sizes, num_filters, num_classes, d_prob, mode, TEXT):
        super(QuantLeNet, self).__init__()
        self.vocab_size = vocab_size
        self.embedding_dim = embedding_dim
        self.kernel_sizes = kernel_sizes
        self.num_filters = num_filters
        self.num_classes = num_classes
        self.d_prob = d_prob
        self.mode = mode
        self.embedding = nn.Embedding(vocab_size, embedding_dim, padding_idx=1)
        self.load_embeddings(TEXT)

   
        self.conv = nn.ModuleList([
            qnn.QuantConv1d(
                in_channels=embedding_dim,
                out_channels=num_filters,
                kernel_size=k, stride=1,
                weight_quant_type=QuantType.INT,
                weight_bit_width=12) for k in kernel_sizes])

      
        self.dropout = nn.Dropout(d_prob)
        self.fc = qnn.QuantLinear(len(kernel_sizes) * num_filters, num_classes,
                                  bias=True,
                                  weight_quant_type=QuantType.INT,
                                  weight_bit_width=32)

    def forward(self, x):
        batch_size, sequence_length = x.shape
        relu = qnn.QuantReLU(
            quant_type=QuantType.INT, bit_width=32, max_val=6)
        sigmoid = qnn.QuantSigmoid(
            quant_type=QuantType.INT, bit_width=32)

        x = self.embedding(x).transpose(1, 2)
        x = [relu(conv(x)) for conv in self.conv]
        x = [F.max_pool1d(c, c.size(-1)).squeeze(dim=-1) for c in x]
        x = torch.cat(x, dim=1)
        x = self.fc(self.dropout(x))
        x = sigmoid(x)
        return x.squeeze()

    def load_embeddings(self, TEXT):
        if 'static' in self.mode:
            self.embedding.weight.data.copy_(TEXT.vocab.vectors)
            if 'non' not in self.mode:
                self.embedding.weight.data.requires_grad = False
                print('Loaded pretrained embeddings, weights are not trainable.')
            else:
                self.embedding.weight.data.requires_grad = True
                print('Loaded pretrained embeddings, weights are trainable.')
        elif self.mode == 'rand':
            print('Randomly initialized embeddings are used.')
        else:
            raise ValueError(
                'Unexpected value of mode. Please choose from static, nonstatic, rand.')`

**Comment by Dakkay** [comment]
Here is the code we used, we took it from the example of Convolutional Model on the brevitas repo.

`class QuantLeNet(Module):
    def __init__(self, vocab_size, embedding_dim, kernel_sizes, num_filters, num_classes, d_prob, mode, TEXT):
        super(QuantLeNet, self).__init__()
        self.vocab_size = vocab_size
        self.embedding_dim = embedding_dim
        self.kernel_sizes = kernel_sizes
        self.num_filters = num_filters
        self.num_classes = num_classes
        self.d_prob = d_prob
        self.mode = mode
        self.embedding = nn.Embedding(vocab_size, embedding_dim, padding_idx=1)
        self.load_embeddings(TEXT)

        # is this ok? if not split list comprehension. it just holds them anyway,
        # no calculations involved
        self.conv = nn.ModuleList([
            qnn.QuantConv1d(
                in_channels=embedding_dim,
                out_channels=num_filters,
                kernel_size=k, stride=1,
                weight_quant_type=QuantType.INT,
                weight_bit_width=8) for k in kernel_sizes])

        # next line is iffy. am i allowed? just remove if i cant, it helps
        # with regularization but maybe we can find another way.
        # at the same time, it *should* be REALLY easy to implement
        self.dropout = nn.Dropout(d_prob)
        self.fc = qnn.QuantLinear(len(kernel_sizes) * num_filters, num_classes,
                                  bias=True,
                                  weight_quant_type=QuantType.INT,
                                  weight_bit_width=8)

    def forward(self, x):
        batch_size, sequence_length = x.shape
        relu = qnn.QuantReLU(
            quant_type=QuantType.INT, bit_width=8, max_val=6)
        sigmoid = qnn.QuantSigmoid(
            quant_type=QuantType.INT, bit_width=8)

        x = self.embedding(x).transpose(1, 2)
        x = [relu(conv(x)) for conv in self.conv]
        x = [F.max_pool1d(c, c.size(-1)).squeeze(dim=-1) for c in x]
        x = torch.cat(x, dim=1)
        x = self.fc(self.dropout(x))
        x = sigmoid(x)
        return x.squeeze()

    def load_embeddings(self, TEXT):
        if 'static' in self.mode:
            self.embedding.weight.data.copy_(TEXT.vocab.vectors)
            if 'non' not in self.mode:
                self.embedding.weight.data.requires_grad = False
                print('Loaded pretrained embeddings, weights are not trainable.')
            else:
                self.embedding.weight.data.requires_grad = True
                print('Loaded pretrained embeddings, weights are trainable.')
        elif self.mode == 'rand':
            print('Randomly initialized embeddings are used.')
        else:
            raise ValueError(
                'Unexpected value of mode. Please choose from static, nonstatic, rand.')`

**Comment by volcacius** [comment]
Hello,

QuantConv1d doesn't support export to FINN yet, and there is no support for embeddings at the moment in either Brevitas or FINN. The sort of models that current can export to FINN out of the box can be found here: https://github.com/Xilinx/brevitas/tree/master/brevitas_examples/bnn_pynq . 

Alessandro
